---
title: "Secondary analysis"
subtitle: "in gene expression studies"
author: "Kontantin Zaitsev"
date: "March 17<sup>th</sup>, 2020"
output:
  xaringan::moon_reader:
    css: ["../../libs/itmo.css", "../../libs/itmo-fonts.css"]
    nature:
      ratio: "16:9"
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
---

class: center, middle

# Secondary analysis

---

## Getting an expression matrix

* Assume bulk gene expression
* Microarray: itensities -> probe/gene symbol mapping -> gene expression matrix
* RNA-seq: raw reads -> alignments -> quantification -> gene expression matrix

---

## Gene expression matrix

* Rows are genes
* Columns are samples
* We assume columns to be somewhat normalized, so gene expression levels are representative across all samples

```
        GSM3703675 GSM3703676 GSM3703677 GSM3703678 GSM3703679 GSM3703680 GSM3703681 GSM3703682
Dhx36     6.868925   7.802883   7.265303   7.711423   7.925366   8.064470   7.782933   7.878214
Arl6ip4   8.610726   8.349129   8.961090   8.863572   8.210400   8.298973   8.612580   8.380108
Tram1     8.117650   7.725020   9.782122   9.205673   8.632618   8.403167  10.311316   9.872407
Mir425    4.124838   4.674299   5.128062   4.936080   4.404071   4.785895   4.397090   4.670425
Pex6      8.295669   8.071793   7.823910   8.110941   7.758333   7.989627   7.529114   7.686020
Nans      8.768137   8.620869   9.358366   9.096089   8.752107   8.696490   9.641165   9.123614
...
```

---

## Once we have an expression matrix


Conceptual analysis steps are the same:

* **Quality controls: PCA + outlier/batch removal if needed**
* Differential expression design
* Performing differential expression
* DE genes: looking for possible biological pathways, transcriptional factors, regulators...

---

## Sources of variance

It is important to identify sources of gene expression variance

* Variation included by design: cell type, treatment, cases vs controls
* Biological sources of variation: cell cycle, sex of mice/donor, cell types present in the sample
* Unwanted variaton: batch effect, donor effect
* Technical variation: microarray variation, sequencing variation

---

## libraries

```{r message=FALSE}

if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
if (!requireNamespace("ggplot2", quietly = TRUE)) install.packages("ggplot2")
if (!requireNamespace("pheatmap", quietly = TRUE)) install.packages("pheatmap")
if (!requireNamespace("RColorBrewer", quietly = TRUE)) install.packages("RColorBrewer")
if (!requireNamespace("ggrepel", quietly = TRUE)) install.packages("ggrepel")
if (!requireNamespace("dplyr", quietly = TRUE)) install.packages("dplyr")

if (!requireNamespace("limma", quietly = TRUE)) BiocManager::install("limma")
if (!requireNamespace("Biobase", quietly = TRUE)) BiocManager::install("Biobase")
if (!requireNamespace("sva", quietly = TRUE)) BiocManager::install("sva")
if (!requireNamespace("fgsea", quietly = TRUE)) BiocManager::install("fgsea")
```


---

## Understanding variance

```{r message=FALSE}
library(Biobase)
library(limma)
library(sva)
library(ggplot2)
library(pheatmap)
library(RColorBrewer)
library(ggrepel)
library(dplyr)
library(fgsea)

blueWhiteRed <- colorRampPalette(c("#3859A8", "#EEEEEE", "#EE2930"))(10)

load("gse129260.Rdata")

```
---

## Understanding variance

```{r message=FALSE}

head(exprs(gse129260))

```
---


## Understanding variance

```{r fig.height=3, fig.fullwidth=T, dev='svg'}

someGenes <- exprs(gse129260)[c("Actb", "Ddx3y", "Il10"), ]
plotData <- t(someGenes)
plotData <- as.data.frame(plotData)
plotData <- cbind(plotData, pData(gse129260))

head(plotData, 4)


```
---

## Understanding variance

```{r fig.show = 'hold', fig.width=3, fig.height=3, fig.fullwidth=T, dev='svg'}
ggplot(plotData, aes(y=Actb)) +
  geom_boxplot() + theme_bw()
ggplot(plotData, aes(x=Cell, y=Actb)) +
  geom_boxplot() + theme_bw()
ggplot(plotData, aes(x=Replicate, y=Actb)) +
  geom_boxplot() + theme_bw()

```
---

## Understanding variance

```{r fig.show = 'hold', fig.width=3, fig.height=3, fig.fullwidth=T, dev='svg'}
ggplot(plotData, aes(y=Il10)) +
  geom_boxplot() + theme_bw()
ggplot(plotData, aes(x=Cell, y=Il10)) +
  geom_boxplot() + theme_bw()
ggplot(plotData, aes(x=Replicate, y=Il10)) +
  geom_boxplot() + theme_bw()
```
---

## Understanding variance

```{r fig.show = 'hold', fig.width=3, fig.height=3, fig.fullwidth=T, dev='svg'}
ggplot(plotData, aes(y=Ddx3y)) +
  geom_boxplot() + theme_bw()
ggplot(plotData, aes(x=Cell, y=Ddx3y)) +
  geom_boxplot() + theme_bw()
ggplot(plotData, aes(x=Replicate, y=Ddx3y)) +
  geom_boxplot() + theme_bw()
```
---

## Heatmaps:

In gene set expression studies we usually use heatmaps to visualize expression levels:

```{r ig.show = 'hold', fig.width=7, fig.height=3, fig.fullwidth=T, dev='svg'}

pheatmap(someGenes, scale="row", color=blueWhiteRed, annotation_col = pData(gse129260), cluster_cols = F)

```

---

## Heatmaps: clustered

```{r ig.show = 'hold', fig.width=7, fig.height=4, fig.fullwidth=T, dev='svg'}

pheatmap(exprs(gse129260), scale="row", color=blueWhiteRed, border_color = NA, kmeans_k = 8,
         annotation_col = pData(gse129260), cluster_cols = F)

```

---

## PCA: exploring the variance

PCA (principal compononent analysis):

* PCA puts samples in new space
* Principal components (new coordinates) explain as much variance as possible in original space
* Samples in original space are far from each other if a lot of genes are differentially expressed between them


---

## PCA: exploring the variance

.center[
<img src="http://www.nlpca.org/fig_pca_principal_component_analysis.png" />
]


---

## PCA: exploring the variance

```{r fig.show = 'hold', fig.width=3, fig.height=3, fig.fullwidth=T, dev='svg'}
pcas <- prcomp(t(exprs(gse129260)), scale. = T)
plotData <- cbind(pcas$x[, 1:2], pData(gse129260))
```

---

## PCA: exploring the variance

```{r fig.show = 'hold', fig.width=5, fig.height=3, fig.fullwidth=T, dev='svg'}
ggplot(plotData, aes(x=PC1, y=PC2, color=Cell, shape=Treatment)) +
  geom_point() + theme_bw() + theme(aspect.ratio = 1)
ggplot(plotData, aes(x=PC1, y=PC2, color=Replicate, shape=Treatment)) +
  geom_point() + theme_bw() + theme(aspect.ratio = 1)
```

---

## PCA: exploring the variance

```{r fig.show = 'hold', fig.width=3, fig.height=3, fig.fullwidth=T, dev='svg'}
rotation <- pcas$rotation
PC1GenesDown <- head(rownames(rotation[order(rotation[, 1]), ]), 10)
PC1GenesUp <- tail(rownames(rotation[order(rotation[, 1]), ]), 10)
print(PC1GenesDown)
print(PC1GenesUp)
```

---

## Heatmaps: PC1 genes

```{r ig.show = 'hold', fig.width=7, fig.height=4, fig.fullwidth=T, dev='svg'}

pheatmap(exprs(gse129260)[c(PC1GenesDown, PC1GenesUp), ], 
         scale="row", color=blueWhiteRed, border_color = NA,
         annotation_col = pData(gse129260), cluster_cols = F)

```

---

## PCA: exploring the variance

```{r fig.show = 'hold', fig.width=3, fig.height=3, fig.fullwidth=T, dev='svg'}
rotation <- pcas$rotation
PC2GenesDown <- head(rownames(rotation[order(rotation[, 2]), ]), 10)
PC2GenesUp <- tail(rownames(rotation[order(rotation[, 2]), ]), 10)
print(PC2GenesDown)
print(PC2GenesUp)
```

---

## Heatmaps: PC2 genes

```{r ig.show = 'hold', fig.width=7, fig.height=4, fig.fullwidth=T, dev='svg'}

pheatmap(exprs(gse129260)[c(PC2GenesDown, PC2GenesUp), ], 
         scale="row", color=blueWhiteRed, border_color = NA,
         annotation_col = pData(gse129260), cluster_cols = F)

```


---

## Can we remove unwanted sources of variance ?

In our experiment we wanted to get variance from Treatment + Cell type

* Can we identify sources of unwanted variance?
* If we know source of variance, can we remove it?

---

## Batch correction: ComBat from SVA

SVA package:

* Allows to identify latent variables
* Allows to remove unwanted variance (ComBat)
* "Adjusting batch effects in microarray expression data using empirical Bayes methods"

<div class="my-footer"><span>Original paper here https://academic.oup.com/biostatistics/article/8/1/118/252073</span></div>

---

## Batch correction: ComBat

```{r }
batch <- pData(gse129260)$Replicate
modcombat <- model.matrix(~1, data=pData(gse129260))
combat_gse129260 = ComBat(dat=exprs(gse129260), batch=batch, mod=modcombat)
```

---

## Batch correction: ComBat

```{r fig.show = 'hold', fig.width=6, fig.height=3, fig.fullwidth=T, dev='svg'}
pcas <- prcomp(t(combat_gse129260), scale. = T)
plotData <- cbind(pcas$x[, 1:2], pData(gse129260))
ggplot(plotData, aes(x=PC1, y=PC2, color=Cell, shape=Treatment)) +
  geom_point() + theme_bw() + theme(aspect.ratio = 1)
```

---
## ComBat: removes batch variance

```{r fig.height=3, fig.fullwidth=T, dev='svg'}

someGenes <- combat_gse129260[c("Actb", "Ddx3y", "Il10"), ]
plotData <- t(someGenes)
plotData <- as.data.frame(plotData)
plotData <- cbind(plotData, pData(gse129260))

head(plotData)


```
---

## ComBat: removes batch variance

```{r fig.show = 'hold', fig.width=3, fig.height=3, fig.fullwidth=T, dev='svg'}
ggplot(plotData, aes(y=Actb)) +
  geom_boxplot() + theme_bw()
ggplot(plotData, aes(x=Cell, y=Actb)) +
  geom_boxplot() + theme_bw()
ggplot(plotData, aes(x=Replicate, y=Actb)) +
  geom_boxplot() + theme_bw()

```
---

## ComBat: removes batch variance

```{r fig.show = 'hold', fig.width=3, fig.height=3, fig.fullwidth=T, dev='svg'}
ggplot(plotData, aes(y=Il10)) +
  geom_boxplot() + theme_bw()
ggplot(plotData, aes(x=Cell, y=Il10)) +
  geom_boxplot() + theme_bw()
ggplot(plotData, aes(x=Replicate, y=Il10)) +
  geom_boxplot() + theme_bw()
```
---

## ComBat: removes batch variance

```{r fig.show = 'hold', fig.width=3, fig.height=3, fig.fullwidth=T, dev='svg'}
ggplot(plotData, aes(y=Ddx3y)) +
  geom_boxplot() + theme_bw()
ggplot(plotData, aes(x=Cell, y=Ddx3y)) +
  geom_boxplot() + theme_bw()
ggplot(plotData, aes(x=Replicate, y=Ddx3y)) +
  geom_boxplot() + theme_bw()
```
---

## Variance

* Varince in transcriptional data comes from both signal and noise
* In ideal scenario the only source of variance is included by your experimental design
* In reality: batch effect, donor effect (super common for human data)
* We can remove unwanted sources of variance if they introduce too much variance

---
## Once we have an expression matrix


Conceptual analysis steps are the same:

* Quality controls: PCA + outlier/batch removal if needed
* **Differential expression design**
* Performing differential expression
* DE genes: looking for possible biological pathways, transcriptional factors, regulators...

---

## Linear models

The most simple linear models are:

$$ y = kx + b $$

* We know both $y$ and $x$ and we try to predict $k$ and $b$
* Usually both $x$ and $y$ are numeric

---

## Linear models: x can be factor

Let's look at expression of gene Il10


<a href="https://www.codecogs.com/eqnedit.php?latex=\LARGE&space;y&space;=&space;k_{pos}&space;x_{pos}&space;&plus;&space;k_{neg}&space;x_{neg}" target="_blank"><img src="https://latex.codecogs.com/svg.latex?\LARGE&space;y&space;=&space;k_{pos}&space;x_{pos}&space;&plus;&space;k_{neg}&space;x_{neg}" title="\LARGE y = k_{pos} x_{pos} + k_{neg} x_{neg}" /></a>

where 
* $x_{pos} = 1$ and $x_{neg} = 0$ if sample is *Il10*-positive 
* $x_{pos} = 0$ and $x_{neg} = 1$ if sample is *Il10*-negative.

---

## Linear models: 

```{r }
model_simple <- model.matrix(~0 + Cell, data = pData(gse129260))
colnames(model_simple) <- c("Negative", "Positive")
model_simple
```

---

## Linear models: 

```{r }
exprs(gse129260)["Il10", ]
linear_fit <- lm.fit(model_simple, exprs(gse129260)["Il10", ])
linear_fit$coef
```

---

## Linear models: 

We can model expression of Il10 gene as:

* $Il10 = 9.178349$ if sample is Il10 negative
* $Il10 = 12.067074$ if sample is Il10 positive

---

## More complicated linear models

* Il10 can be well-modeled with single factor variable
* Let's look at S100a10 gene

```{r fig.show = 'hold', fig.width=6, fig.height=2, dev='svg'}
pheatmap(exprs(gse129260)[c("Il10", "S100a10"), ], scale="row", cluster_cols = F, cluster_rows = F, color=blueWhiteRed, annotation_col = pData(gse129260))
```

---

## Linear models: Treatment

```{r }
treatment_model <- model.matrix(~0 + Treatment, data = pData(gse129260))
colnames(treatment_model) <- c("aCD-40", "LPS")
treatment_model
```

---

## Linear models: Treatment + Replicate

```{r }
treatment_rep_model <- model.matrix(~0 + Treatment + Replicate, data = pData(gse129260))
colnames(treatment_rep_model) <- c("aCD-40", "LPS", "Rep2")
treatment_rep_model
```

---

## Linear models: 

```{r }
exprs(gse129260)["S100a10", ]
linear_fit <- lm.fit(treatment_rep_model, exprs(gse129260)["S100a10", ])
linear_fit$coef
```

---

## Linear models: Treatment + Replicate

* Including several variables in the design allows us to calculate effects for each variable
* First variable is usually a target for differential expression
* Only the first variable will have both 0/1 effect calculated

---

## Linear models: full model

```{r }
full_model <- model.matrix(~0 + Treatment + Cell + Replicate, data = pData(gse129260))
colnames(full_model) <- c("aCD-40", "LPS", "Il10pos", "Rep2")
full_model
```

---

## Linear models: full model

```{r }
linear_fit <- lm.fit(full_model, exprs(gse129260)["Il10", ])
linear_fit$coef

linear_fit <- lm.fit(full_model, exprs(gse129260)["S100a10", ])
linear_fit$coef
```

---

## Linear models: full model with 1

```{r }
full_model <- model.matrix(~1 + Treatment + Cell + Replicate, data = pData(gse129260))
colnames(full_model) <- c("Intercept", "LPS", "Il10pos", "Rep2")
full_model
```

---

## Linear models: full model with 1

```{r }
linear_fit <- lm.fit(full_model, exprs(gse129260)["Il10", ])
linear_fit$coef

linear_fit <- lm.fit(full_model, exprs(gse129260)["S100a10", ])
linear_fit$coef
```

---

## Linear models

* Linear models are usefull for calculating effects of variables
* ` ~ 0 + Annotation1 + Annotation2 ... ` will calculate means for both factors in Annotation1, and calculate effects for other annotations (excluding Annotation1)
* ` ~ 1 + Annotation1 + Annotation2 ... ` will calculate means for one the Annotation1 factors, and calculate effects for all annotations (including remaining factor in Annotation1)


---

## Once we have an expression matrix


Conceptual analysis steps are the same:

* Quality controls: PCA + outlier/batch removal if needed
* Differential expression design
* **Performing differential expression**
* DE genes: looking for possible biological pathways, transcriptional factors, regulators...

---

## Performing DE (differential expression)

* Performing DE is usually much easier than designing proper DE :)
* We will use limma in this example
* Let's check which design get more results


---

## Performing limma

```{r }

cell_full_model <- model.matrix(~0 + Cell + Treatment + Replicate, data=pData(gse129260))
colnames(cell_full_model) <- c("il10neg", "il10pos", "LPS", "rep2")

fit <- lmFit(gse129260, cell_full_model)

fit2 <- contrasts.fit(fit, makeContrasts(il10pos - il10neg, levels=cell_full_model))
fit2 <- eBayes(fit2, trend = T)

de <- topTable(fit2, adjust.method="BH", number=Inf, sort.by = "P")

```

---

## Differential expression

```{r }
head(de)
```

---

## Volcano plot

* Volcano plot is the usual way to display DE results
* X axis is log fold change showing the direction of the change
* Y axis is -log10(p adjusted) - showing the significane of DE


---

## DE: Volcano plot

```{r fig.show = 'hold', fig.width=5, fig.height=3.5, fig.fullwidth=T, dev='svg'}

ggplot(de, aes(x=logFC, y=-log10(adj.P.Val), color=adj.P.Val < 0.05)) +
  geom_point() + theme_bw()


```

---

## DE: Volcano plot (fancy)

```{r fig.show = 'hold', fig.width=8, fig.height=3.5, fig.fullwidth=T, dev='svg'}

ggplot(de, aes(x=logFC, y=-log10(adj.P.Val), color=adj.P.Val < 0.05)) +
  geom_point() + theme_bw() + scale_color_manual(values=c("black", "red")) +
  geom_text_repel(data=de %>% dplyr::filter(adj.P.Val < 0.01), aes(label=Gene.symbol, color=NULL))


```

---

## Performing limma: bad model

```{r }

cell_bad_model <- model.matrix(~0 + Cell, data=pData(gse129260))
colnames(cell_bad_model) <- c("il10neg", "il10pos")

fit_bad <- lmFit(gse129260, cell_bad_model)

fit_bad2 <- contrasts.fit(fit_bad, makeContrasts(il10pos - il10neg, levels=cell_bad_model))
fit_bad2 <- eBayes(fit_bad2, trend = T)

de_bad <- topTable(fit_bad2, adjust.method="BH", number=Inf, sort.by = "P")
```

---

## Comparing results

```{r }

de %>% filter(adj.P.Val < 0.05) %>% count()
de_bad %>% filter(adj.P.Val < 0.05) %>% count()

```

---

## Differential expression

* Good design empowers you to find differences that you want to find in your data
* Design with a smaller number of variables ignores additional information and assumes samples in a group to be the same
* Design with a smaller number of variables is still something you might want to do

---

## Once we have an expression matrix


Conceptual analysis steps are the same:

* Quality controls: PCA + outlier/batch removal if needed
* Differential expression design
* Performing differential expression
* **DE genes: looking for possible biological pathways, transcriptional factors, regulators...**

---

## Pathway enrichment: 

* In most cases gene expression changes are not coming one gene by one
* Genes that are changed are regulated by biological processes (pathways)
* We have bunch of databases that describe gene sets: sets of genes that regulate or regulated by biological process
* Technically speaking, for us pathway is just a set of genes

---

## Pathway enrichment: 

* Pathways
* GO terms
* Targets of transcriptional factors
* Gene sets produced by other datasets

---

## Pathway enrichment: 

Let's didcuss terms firts:

* Universe: genes that are expressed in the dataset (in our case size of the universe 12000):

<script type="math/tex; mode=display">
						U = \{g_1, g_2, \dots, g_n\}, \ \  |U| = n \approx 12000
</script>

* We have $N$ pathways: 

<script type="math/tex; mode=display">
						P_i = \{ g_{i, 1}, g_{i, 2}, \dots, g_{i, m_i} \} \\
						|P_i| = m_i, \ \ \ g_{i, j} \in U
</script>


---

## Pathway enrichment:

* We have results of our DE
* "Pathway behaves non-random" = "Genes from pathways are not changing randomly"
* We want to identify pathways that behave non-random in our DE results

---

## Simple implementation: exact Fisher test

Exact Fisher test (or hypergeometric test)

* We only choose significantly expressed genes
* We test overlaps of these genes with pathway
* Identify if overlap is random

---

## Loading kegg Pathways

```{r }
load("keggSymbolMouse.rdata")
upRegulatedGenes <- de %>% filter(adj.P.Val < 0.05 & logFC > 0) %>% pull("Gene.symbol")
length(upRegulatedGenes)

randomGeneSet <- keggSymbolMouse[["Cardiac muscle contraction - Mus musculus (mouse)"]]
randomGeneSet <- randomGeneSet[randomGeneSet %in% rownames(de)]
length(randomGeneSet)

length(intersect(randomGeneSet, upRegulatedGenes))

```
---

## Hypergeometric test

https://en.wikipedia.org/wiki/Hypergeometric_distribution

* N = 12000: total number of genes (TOTAL)
* K = 41: number of genes in pathway (SUCCESSES)
* n = 113: number of DE genes (DRAWS)
* k = 1: overlap (SUCCESSFUL DRAWS)

Null hypothesis -- genes are drawn from 12000 genes at random with respect to the pathway

---

## Running hypergeometric test

```{r }

N <- nrow(de)
K <- length(randomGeneSet)
n <- length(upRegulatedGenes)
k <- length(intersect(upRegulatedGenes, randomGeneSet))
phyper(k - 1, K, N - K, n, lower.tail = F)
```
---

## Non-random set

```{r }

nonRandomGeneSet <- keggSymbolMouse[["Cytokine-cytokine receptor interaction - Mus musculus (mouse)"]]
nonRandomGeneSet <- nonRandomGeneSet[nonRandomGeneSet %in% rownames(de)]


N <- nrow(de)
K <- length(nonRandomGeneSet)
n <- length(upRegulatedGenes)
k <- length(intersect(upRegulatedGenes, nonRandomGeneSet))
print(c(N, K, n, k))
phyper(k - 1, K, N - K, n, lower.tail = F)
```

---

## Hypergeometric tests

* Require you to define gene set to test:
  * Setting arbitrary threshold (< 0.01 or < 0.05)
  * Only work with decent amount of genes (hard to calculate overlaps for 20 genes)
* Very robust with large number of genes
* Many databases offer you hypergeometric test with FDR correction (multiple pathways tests)
* http://software.broadinstitute.org/gsea/msigdb/annotate.jsp (you can enter my email for now: zayats1812@mail.ru)

---

## Hypergeometric tests

* http://software.broadinstitute.org/gsea/msigdb/annotate.jsp (you can enter my email for now: zayats1812@mail.ru)

```{r }
cat(upRegulatedGenes)
```

---

## GSEA (gene-set enrichment analysis)

* GSEA uses information about all genes in DE to score pathways
* Genes are ranked by their difference in DE (usually by t statistic)
* We try to identify pathways for which genes are distributed at random

---

## GSEA (gene-set enrichment analysis)

We will use fgsea package (the guys in our lab are amazing):

```{r fig.show = 'hold', fig.width=8, fig.height=3, fig.fullwidth=T, dev='svg'}
stats <- de$t
names(stats) <- de$Gene.symbol
plotEnrichment(randomGeneSet, stats)

```

---

## GSEA (gene-set enrichment analysis)


```{r fig.show = 'hold', fig.width=8, fig.height=3.5, fig.fullwidth=T, dev='svg'}

plotEnrichment(nonRandomGeneSet, stats)

```

---

## FGSEA

* Based on enrichment score we can calculate p value for each pathway
* fgsea allows us to do it quickly (f for FAST) for all the pathways in the same time

---

## FGSEA

```{r }
fgseaResults <- fgseaMultilevel(keggSymbolMouse, stats, minSize = 15, maxSize = 500)
head(fgseaResults, 3)

```

---

## FGSEA

```{r fig.show = 'hold', fig.width=10, fig.height=3.5, fig.fullwidth=T, dev='svg'}
topPathwaysUp <- fgseaResults[ES > 0, ][head(order(pval), n=5), pathway]
topPathwaysDown <- fgseaResults[ES < 0, ][head(order(pval), n=5), pathway]
topPathways <- c(topPathwaysUp, rev(topPathwaysDown))
```

---

## FGSEA

```{r fig.show = 'hold', fig.width=10, fig.height=4, fig.fullwidth=T, dev='svg'}
plotGseaTable(keggSymbolMouse[topPathways], stats, fgseaResults, gseaParam = 0.5)
```

---
## FGSEA:

* Does not require a priori threshold to define DE genes
* Can detect a lot of small changes
* Detects up/down pathways in the same time

---

## Conclusion

Secondary analysis is about making sense of the data:
* Making sense of the variance
* Making sense of differential expression design
* Making sense of differential expression results

Once we have differential expression we can try to guess what's going on:
* Biological pathways
* TF targets

---
class: center, middle

# Questions?