<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Introduction into single cell RNA-seq</title>
    <meta charset="utf-8" />
    <meta name="author" content="Konstantin Zaitsev" />
    <link rel="stylesheet" href="libs/itmo.css" type="text/css" />
    <link rel="stylesheet" href="libs/itmo-fonts.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Introduction into single cell RNA-seq
### Konstantin Zaitsev
### December 3<sup>rd</sup>, 2020. Tomsk / Saint-Petersburg

---


class: center, middle

# Single-cell RNA-seq (scRNA-seq)

---

## Cell is the fundamental unit

* Microscopy
* FACS (fluorescence activated cell sorting)/ CyTOF (Cytometry by Time Of Flight)
* scRNA-seq (single-cell RNA-seq)
* Single cell genomics and epigenetics

---

## Single cell RNA-seq

.pull-left[
* RNA-seq is a snapshot of what is happening in the sample
* Sample consists of many different cells and cell types
* Single-cell RNA-seq - thousand of individual snapshots of many cells to capture the whole picture
]
.pull-right[
  .center[&lt;img src="islets.png" width="60%"&gt;]
]

---

## Why single-cell RNA-seq


Heterogeneous populations:
* New cell subpopulations discovery
* Comparison of similar cell subpopulations
* Marker selection for cell subpopulations

Homogeneous populations:
* Understanding heterogeneity
* Cellular states and cellular processes

Tracking of cell differentiation

---

## Smart-seq2

.center[&lt;img src="smartseq2.png" width="80%"&gt;]

&lt;div class="my-footer"&gt;&lt;span&gt;Paper: https://www.nature.com/articles/nprot.2014.006&lt;/span&gt;&lt;/div&gt; 


---

## Single-cell RNA-seq of myeloid cells

.pull-left[
* Villani, Satija et al
* Science, 2017
* 1152 cells
]

.pull-right[
.center[&lt;img src="myeloid_fix.png" width="100%"&gt;]
]


&lt;div class="my-footer"&gt;&lt;span&gt;Paper: https://science.sciencemag.org/content/356/6335/eaah4573&lt;/span&gt;&lt;/div&gt; 

---

## Single-cell RNA-seq of myeloid cells

.center[&lt;img src="design_myeloid.png" width="100%"&gt;]


&lt;div class="my-footer"&gt;&lt;span&gt;Paper: https://science.sciencemag.org/content/356/6335/eaah4573&lt;/span&gt;&lt;/div&gt; 


---

## Single-cell RNA-seq of melanoma

.pull-left[
* Tirosh, Izar et al
* Science, 2016
* 4645 cells
]

.pull-right[
&lt;img src="sc_cancer.png" width="110%"&gt;
]

&lt;div class="my-footer"&gt;&lt;span&gt;Paper: https://science.sciencemag.org/content/352/6282/189.long&lt;/span&gt;&lt;/div&gt; 

---

## Drop-seq: Cell, 2015

.center[
&lt;img src="dropseq.png" width="60%"&gt;
]

&lt;div class="my-footer"&gt;&lt;span&gt;Paper: https://www.cell.com/abstract/S0092-8674(15)00549-8&lt;/span&gt;&lt;/div&gt; 

---

## 10x chromium machine: NComms, 2017

.center[
&lt;img src="10x.png" width="65%"&gt;
]

&lt;div class="my-footer"&gt;&lt;span&gt;Paper: https://www.nature.com/articles/ncomms14049&lt;/span&gt;&lt;/div&gt; 

---

## Seq-Well: NMeth, 2017

.pull-left[
&lt;img src="seqwell_head.png" width="100%"&gt;
]

.pull-right[
&lt;img src="seqwell_design.png" width="100%"&gt;
]

&lt;div class="my-footer"&gt;&lt;span&gt;Paper: https://www.nature.com/articles/nmeth.4179&lt;/span&gt;&lt;/div&gt; 

---

## Microwell-Seq: Cell, 2018

.center[
&lt;img src="microwellseq.png" width="60%"&gt;
]

&lt;div class="my-footer"&gt;&lt;span&gt;Paper: https://www.ncbi.nlm.nih.gov/pubmed/29474909&lt;/span&gt;&lt;/div&gt; 

---

## Growth of single-cell technologies

.center[
&lt;img src="cells.png" width="105%"&gt;
]

&lt;div class="my-footer"&gt;&lt;span&gt;Human cell atlas white paper: https://arxiv.org/pdf/1810.05192.pdf&lt;/span&gt;&lt;/div&gt; 

---

## Growth of single-cell technologies

.center[
&lt;img src="papers.png" width="90%"&gt;
]

&lt;div class="my-footer"&gt;&lt;span&gt;Based on pubmed results&lt;/span&gt;&lt;/div&gt; 

---

## Main challenges

* **How to amplify and sequence small number of RNA (typical mammalian cell has only 200 000 mRNA molecules)?**
* How to isolate cells?
* How to work with big number of cells?

---


## Low library complexity

* Small numbers of mRNA molecules yield low complexity cDNA library
* cDNA molecules to be amplified by PCR
* We don’t want to sequence tons of PCR duplicates

.center[
&lt;img src="pcr.png" width="80%"&gt;
]

---

## Template-switching PCR

.pull-left[
* Switching Mechanism At the 5' end of RNA Template (SMART)
* Robust for low input libraries
]

.pull-right[
&lt;img src="template_switch.png" width="90%"&gt;
]

---

## Drop-seq schematics

.center[
&lt;img src="scheme_1.png" width="95%"&gt;
]

&lt;div class="my-footer"&gt;&lt;span&gt;Paper: https://www.cell.com/abstract/S0092-8674(15)00549-8&lt;/span&gt;&lt;/div&gt;
---

## Drop-seq microfluidics

.pull-left[
&lt;img src="microfluidics1.jpg" width="90%"&gt;
]

.pull-right[
&lt;img src="microfluidics2.jpg.png" width="110%"&gt;
]

&lt;div class="my-footer"&gt;&lt;span&gt;Paper: https://www.cell.com/abstract/S0092-8674(15)00549-8&lt;/span&gt;&lt;/div&gt;

---

## Drop-seq microfluidics

.center[
&lt;img src="dropseq_gif.gif" width="100%"&gt;
]

---

## Drop-seq schematics

.center[
&lt;img src="scheme_2.png" width="95%"&gt;
]

&lt;div class="my-footer"&gt;&lt;span&gt;Paper: https://www.cell.com/abstract/S0092-8674(15)00549-8&lt;/span&gt;&lt;/div&gt;

---

## Barcoding

.center[
&lt;img src="barcoded_beads-01.png" width="100%"&gt;
]

---

## Barcoding

.center[
&lt;img src="barcoded_beads-03.png" width="100%"&gt;
]

---

## Barcoding

.center[
&lt;img src="barcoded_beads-02.png" width="100%"&gt;
]

---

## Barcoding

.center[
&lt;img src="barcoded_beads-04.png" width="100%"&gt;
]

---

## Drop-seq schematics

.center[
&lt;img src="scheme_3.png" width="95%"&gt;
]

&lt;div class="my-footer"&gt;&lt;span&gt;Paper: https://www.cell.com/abstract/S0092-8674(15)00549-8&lt;/span&gt;&lt;/div&gt;

---

## Sequencing

.center[
&lt;img src="umis1.png" width="100%"&gt;
]
---

## Sequencing

.center[
&lt;img src="umis2.png" width="100%"&gt;
]
---

## Sequencing

.center[
&lt;img src="umis3.png" width="100%"&gt;
]
---

## Sequencing

.center[
&lt;img src="umis4.png" width="100%"&gt;
]
---

## Sequencing

.center[
&lt;img src="umis5.png" width="100%"&gt;
]
---

## Sequencing

.center[
&lt;img src="umis6.png" width="100%"&gt;
]
---

## Sequencing

.center[
&lt;img src="umis7.png" width="100%"&gt;
]
---

## Sequencing

.center[
&lt;img src="umis8.png" width="100%"&gt;
]

---

## All questions were addressed

* Low input mRNA – template switching PCR
* Cell isolation – microfluidics
* Read identification – Cell barcodes
* Dealing with PCR duplicates – Cell/UMI barcodes

---

## All questions were addressed

* Low input mRNA – template switching PCR
* **Cell isolation – microfluidics (might vary from technology to technology)**
* **Barcoding with beads proved to be very effective**

---

## Dataset for today:

* https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.2/5k_pbmc_v3_nextgem

.center[
&lt;img src="5k_v3.png" width="70%"&gt;
]

---

## Let's have a loot at the summary

* http://cf.10xgenomics.com/samples/cell-exp/3.0.2/5k_pbmc_v3_nextgem/5k_pbmc_v3_nextgem_web_summary.html

---

## Understaing what's noise

.center[
&lt;img src="filtering_procedure.png" width="60%"&gt;
]

---

## Understaing what's noise

There is cell-free RNA in the cellular suspension that will be captured in empty droplets (with beads), and we must distinguish cells from empty droplets:

* We can set a hard threshold on UMI to filter noise from actual cells
* We can calculate noise signature and test every cell against this signature (emptyDrops, https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1662-y)

---

## Basic steps to analysis of scRNA-seq

* Filtering out “bad” barcodes
* Normalizing expression levels: (scaling and log2 normalizing)
* PCA
* Visualization (tSNE or UMAP)
* Clustering
* Cellular subset annotation

---

## Our setup

* Address is the same https://ctlab.itmo.ru/rstudio-sbNN/
* Folder scrna-seq

.center[
&lt;img src="other_screen_shots/rstudio1.png" width="60%"&gt;
]

---

## Our setup

* Open the folder
* Set this as a working directory
* Open file scrna-seq-intro.R

.center[
&lt;img src="other_screen_shots/rstudio2.png" width="60%"&gt;
]

---

## Our setup

* You will see chunks of code appearing on my slides
* Follow along!
* Select the code and press Ctrl + Enter to run the chunk

---

## Loading all the librarires 


```r
library(Seurat)
library(Matrix)
library(MAST)
library(ggplot2)
library(dplyr)
```

---

## Loading the data 


```r
data &lt;- Read10X("filtered_feature_bc_matrix/")
dim(data)
```

```
## [1] 33538  5155
```

Count matrix is large:

* 33538 genes 
* 11769 cells



---

## UMI distribution


```r
plotData &lt;- data.frame(
  umis &lt;- colSums(data)
)
ggplot(data=plotData, aes(x=umis)) +
  geom_histogram() + theme_bw()
```

![](scrna-seq-intro_files/figure-html/unnamed-chunk-3-1.svg)&lt;!-- --&gt;

---

## Filtering (genes and barcodes)


```r
seurat &lt;- CreateSeuratObject(data, min.cells = 10, min.features = 10)
dim(seurat)
```

```
## [1] 16022  5155
```

17527 of 33538 genes were filtered (detected in `\(\leq 10\)` cells)

---

## Filtering (genes and barcodes)


```r
FeatureScatter(seurat, "nCount_RNA", "nFeature_RNA") + scale_x_log10() + scale_y_log10()
```

![](scrna-seq-intro_files/figure-html/unnamed-chunk-5-1.png)&lt;!-- --&gt;

---

## Filtering (genes and barcodes)


```r
seurat[["percent.mt"]] &lt;- PercentageFeatureSet(seurat, pattern = "^MT-")
FeatureScatter(seurat, "nCount_RNA", "percent.mt") + scale_x_log10()
FeatureScatter(seurat, "nFeature_RNA", "percent.mt") + scale_x_log10()
```

![](scrna-seq-intro_files/figure-html/unnamed-chunk-6-1.png)![](scrna-seq-intro_files/figure-html/unnamed-chunk-6-2.png)

---

## Filtering (genes and barcodes)


```r
seurat &lt;- subset(seurat, subset = nFeature_RNA &gt; 300 &amp; percent.mt &lt; 25)
dim(seurat)
```

```
## [1] 16022  4714
```

---

## Normalization (old way)


```r
## seurat &lt;- NormalizeData(seurat, normalization.method = "LogNormalize", scale.factor = 10000)
## seurat &lt;- FindVariableFeatures(seurat, selection.method = "vst", nfeatures = 2000)
## seurat &lt;- ScaleData(seurat)
```

* Scaling expression to 10 000 UMIs (instead of million in RPM)
* Finding features with high variance (features above mean/variance trend)
* Scaling data (for PCA and so on)

---

## Normalization (better way)


```r
seurat &lt;- SCTransform(seurat, vars.to.regress = "percent.mt", verbose = FALSE)
```

* SCTransform does all of these things in one command

---

## SCTransform: 

.center[
&lt;img src="sctransform.png" width="80%"&gt;
]

&lt;div class="my-footer"&gt;&lt;span&gt;Paper: https://www.biorxiv.org/content/biorxiv/early/2019/03/18/576827.full.pdf &lt;/span&gt;&lt;/div&gt;

---
## Variable genes

```r
VariableFeaturePlot(seurat) + scale_y_log10()
```

![](scrna-seq-intro_files/figure-html/unnamed-chunk-10-1.png)&lt;!-- --&gt;

---
## Variable genes

```r
top10_variable_genes &lt;- head(VariableFeatures(seurat), 10)
VariableFeaturePlot(seurat) %&gt;% 
  LabelPoints(points = top10_variable_genes, repel = TRUE) +
  scale_y_log10()
```

![](scrna-seq-intro_files/figure-html/unnamed-chunk-11-1.png)&lt;!-- --&gt;

---

## Basic steps to analysis of scRNA-seq

* Filtering out “bad” barcodes
* Normalizing expression levels: (scaling and log2 normalizing)
* **PCA**
* Visualization (tSNE or UMAP)
* Clustering
* Cellular subset annotation

---
## High-dimensionality of scRNA-seq

Initially matrix is very large in size, this causes different kind of issues:

* Some algorithms are just slow when performed in this high-dimensionality data
* Curse of dimensionality

We usually take several steps to reduce dims before creating 2d clustered map of our dataset

* Keeping only variable genes (since those introduce variance to the dataset)
* PCA will reduce dimensionality to 20-30 first components

---

## PCA


```r
seurat &lt;- RunPCA(seurat, verbose = FALSE)
ElbowPlot(seurat, ndims = 50)
```

![](scrna-seq-intro_files/figure-html/unnamed-chunk-12-1.svg)&lt;!-- --&gt;

---

## Basic steps to analysis of scRNA-seq

* Filtering out “bad” barcodes
* Normalizing expression levels: (scaling and log2 normalizing)
* PCA
* **Visualization (tSNE or UMAP)**
* Clustering
* Cellular subset annotation

---

## Visualization

Both tSNE and UMAP will put our data after PCA into a 2D plane:
* Dots that are close to each other are cells that are transcriptionally similar to each other
* Dots that are far from each other are cells that are trasncriptionally different from each other

Lets look at dimensionality of each cell:
* Variable genes only: 30k+ -&gt; 2-3k of variable genes
* PCA: 2-3k of variable genes -&gt; 20-30 principal components
* tSNE or UMAP: 20-30 PCs -&gt; 2d or 3d plots

---

## tSNE


```r
seurat &lt;- RunTSNE(seurat, dims=1:20)
DimPlot(seurat, reduction = "tsne") + NoLegend()
```

![](scrna-seq-intro_files/figure-html/unnamed-chunk-13-1.png)&lt;!-- --&gt;

&lt;div class="my-footer"&gt;&lt;span&gt;Paper: http://www.jmlr.org/papers/volume9/vandermaaten08a/vandermaaten08a.pdf &lt;/span&gt;&lt;/div&gt;

---

## tSNE

tSNE - "t-Distributed Stochastic Neighbor Embedding (t-SNE) is a technique for dimensionality reduction that is particularly well suited for the visualization of high-dimensional datasets"

* We can calculate distances in original space, and then calculate conditional probabilities `\(p_{i| j}\)` that point `\(i\)` would choose point `\(j\)` as a neighbor. `\(p_{i| j}\)`  are proportional to "distances" from `\(i\)` to all other points (actually probability density around point `\(i\)`, but it doesn't matter here).
* Once all `\(p_{i, j}\)` are calculated in original space we try to find such 2d/3d space that would have similar probabilities

&lt;div class="my-footer"&gt;&lt;span&gt;Paper: http://www.jmlr.org/papers/volume9/vandermaaten08a/vandermaaten08a.pdf &lt;/span&gt;&lt;/div&gt;

---

## UMAP


```r
seurat &lt;- RunUMAP(seurat, dims=1:20)
DimPlot(seurat, reduction = "umap") + NoLegend()
```

![](scrna-seq-intro_files/figure-html/unnamed-chunk-14-1.png)&lt;!-- --&gt;

&lt;div class="my-footer"&gt;&lt;span&gt;Paper: https://arxiv.org/pdf/1802.03426.pdf &lt;/span&gt;&lt;/div&gt;

---

## UMAP

.center[
&lt;img src="umap.png" width="80%"&gt;
]

&lt;div class="my-footer"&gt;&lt;span&gt;Paper: https://arxiv.org/pdf/1802.03426.pdf &lt;/span&gt;&lt;/div&gt;

---

## UMAP

https://pair-code.github.io/understanding-umap/

---

## Comparing two


```r
DimPlot(seurat, reduction = "tsne") + NoLegend()
DimPlot(seurat, reduction = "umap") + NoLegend()
```

![](scrna-seq-intro_files/figure-html/unnamed-chunk-15-1.png)![](scrna-seq-intro_files/figure-html/unnamed-chunk-15-2.png)

---

## Basic steps to analysis of scRNA-seq

* Filtering out “bad” barcodes
* Normalizing expression levels: (scaling and log2 normalizing)
* PCA
* Visualization (tSNE or UMAP)
* **Clustering**
* Cellular subset annotation

---

## Clustering and annotation

Clustering:
* Graph-based clustering (preferred)
* K-means 

Annotation:
* First, check known markers
* For each cluster, perform differential expression: cluster against all others
* Top DE genes expected to be highly distinctive marker genes

---

## Clustering

.pull-left[
* Instead of defining clusters based on distance we first find “shared nearest neighbors"
* Cells that have a lot of neighbors in common, most likely “live in the same neighborhood”
* Algorithm is trying to find such neighborhoods
]

.pull-right[
  .center[&lt;img src="snn-cliq.png" width="80%"&gt;]
]

&lt;div class="my-footer"&gt;&lt;span&gt;Paper: https://academic.oup.com/bioinformatics/article/31/12/1974/214505 &lt;/span&gt;&lt;/div&gt;

---

## Clustering


```r
seurat &lt;- FindNeighbors(seurat, dims = 1:20, verbose = FALSE)
seurat &lt;- FindClusters(seurat, resolution=0.6, verbose = FALSE)
DimPlot(seurat, reduction = "tsne", label = TRUE) + NoLegend()
DimPlot(seurat, reduction = "umap", label = TRUE) + NoLegend()
```

![](scrna-seq-intro_files/figure-html/unnamed-chunk-16-1.png)![](scrna-seq-intro_files/figure-html/unnamed-chunk-16-2.png)

---
## PBMC

.center[&lt;img src="pbmc.png" width="80%"&gt;]

---

## Annotation
.small[
Known markers: CD14, CD79A, CD3D are known markers of Monocytes, B cells and T cells respectively


```r
FeaturePlot(seurat, c("CD14", "CD79A", "CD3D"), cols=c("grey", "red"), reduction="umap", ncol=3)
```

![](scrna-seq-intro_files/figure-html/unnamed-chunk-17-1.png)&lt;!-- --&gt;
]


---

## Annotation

We can run DE to identify markers automatically using MAST test


```r
# max cells per ident is only seed to speed up the whole thing
allMarkers &lt;- FindAllMarkers(seurat, max.cells.per.ident = 100, test.use = "MAST", only.pos = T)
goodMarkers &lt;- allMarkers %&gt;% group_by(cluster) %&gt;% top_n(n = 1, wt = avg_logFC) %&gt;% pull(gene)
goodMarkers
```

```
##  [1] "LEF1"     "IL7R"     "S100A9"   "CCL5"     "CCR7"     "AIF1"    
##  [7] "IGKC"     "GNLY"     "IGLC2"    "IGKC"     "GNLY"     "CD8B"    
## [13] "INTS6"    "HLA-DQA1" "TCF4"     "PPBP"
```

---

## Ways to show expression
.small[
We usually either show expression **on top of reduction plot**, or show a violin plot for expression


```r
FeaturePlot(seurat, goodMarkers[1:3], cols=c("grey", "red"), reduction="umap", ncol=3)
```

![](scrna-seq-intro_files/figure-html/unnamed-chunk-19-1.png)&lt;!-- --&gt;
]
---

## Ways to show expression
.small[
We usually either show expression of top of reduction plot, or show a **violin plot** for expression


```r
VlnPlot(seurat, goodMarkers[1:3], pt.size = 0.1)
```

![](scrna-seq-intro_files/figure-html/unnamed-chunk-20-1.png)&lt;!-- --&gt;
]
---
## Save the Seurat object for later


```r
saveRDS(seurat, file="blood_seurat.rds")
```

---
## Communication is important

* When clusters are found we want to identify which cell subsets are presented, to “annotate” them
* If you are a bioinformatician and you are a single-cell RNA-seq dataset that have been designed/done by you, this is a perfect to time to go and talk to a biologist who performed/designed the experiment
* If you are a biologist who designed/performed single-cell RNA-seq experiment, chances are, you know all cellular subsets and markers better than almost anyone else
* This is where you communicate and try to make sense of the data

---

## Latest things: 5'

* 5’ scRNA-seq is different from 3’ scRNA-seq
* 5’ scRNA-seq allows us to get TCR and IG repertoires of the cell

---

## 3' sequencing

.center[&lt;img src="3prime.png" width="80%"&gt;]

---

## 5' sequencing

.center[&lt;img src="5prime.png" width="80%"&gt;]

---

## Surface protein expression

.center[&lt;img src="epitope1.png" width="50%"&gt;]
.center[&lt;img src="epitope2.png" width="50%"&gt;]

---

## Surface protein expression

.center[&lt;img src="epitopes.png" width="60%"&gt;]

---

## Questions?
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"ratio": "16:9",
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
